File: apps/android/app/src/main/java/com/tvm/privateclient/TvmLiveReflector.kt
package com.tvm.privateclient

import java.net.HttpURLConnection
import java.net.URI
import java.net.URL
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap

/**
 * Live TV on the phone goes through the same kind of proxy as desktop Core:
 * the player is handed `/api/live/proxy/<token>`, never the provider address.
 */
class TvmLiveReflector {
    private val targets = ConcurrentHashMap<String, String>()

    fun publish(url: String): String {
        val token = UUID.randomUUID().toString().replace("-", "")
        targets[token] = url
        return "/api/live/proxy/$token"
    }

    fun serve(token: String, method: String): HttpReply {
        val upstream = targets[token] ?: return HttpReply.json(404, Json.obj("error" to "not_found"))
        return try {
            val connection = (URL(upstream).openConnection() as HttpURLConnection).apply {
                instanceFollowRedirects = true
                connectTimeout = 15_000
                readTimeout = 20_000
                requestMethod = if (method == "HEAD") "HEAD" else "GET"
                setRequestProperty("User-Agent", "VLC/3.0.21 LibVLC/3.0.21")
            }
            val status = connection.responseCode
            if (status !in 200..299 && status != 206) {
                return HttpReply.json(502, Json.obj("error" to "upstream-$status"))
            }
            val type = connection.contentType ?: ""
            val body = if (method == "HEAD") ByteArray(0) else connection.inputStream.use { it.readBytes() }
            if (looksLikePlaylist(upstream, type, body)) {
                val text = body.toString(Charsets.UTF_8)
                val rewritten = rewrite(text, upstream)
                return HttpReply.bytes(200, "application/vnd.apple.mpegurl", rewritten.toByteArray(Charsets.UTF_8))
            }
            HttpReply.bytes(if (method == "HEAD") 200 else 200, mediaType(upstream, type), body)
        } catch (_: Exception) {
            HttpReply.json(502, Json.obj("error" to "unreachable"))
        }
    }

    private fun looksLikePlaylist(url: String, type: String, body: ByteArray): Boolean {
        if (type.contains("mpegurl", ignoreCase = true)) return true
        if (url.substringBefore('?').endsWith(".m3u8", ignoreCase = true)) return true
        return body.isNotEmpty() && body[0] == '#'.code.toByte()
    }

    private fun mediaType(url: String, declared: String): String {
        val mime = declared.substringBefore(';').trim()
        if (mime.contains("mp2t", ignoreCase = true)) return "video/mp2t"
        if (mime.isNotEmpty() && !mime.contains("octet-stream", ignoreCase = true)) return mime
        val path = url.substringBefore('?').lowercase()
        if (path.endsWith(".ts") || path.endsWith(".m2ts")) return "video/mp2t"
        if (path.endsWith(".m3u8")) return "application/vnd.apple.mpegurl"
        return "video/mp4"
    }

    private fun rewrite(text: String, base: String): String {
        val origin = runCatching { URI(base) }.getOrNull() ?: return text
        return text.split("\n").joinToString("\n") { line ->
            if (line.startsWith("#")) rewriteAttributes(line, origin)
            else {
                val trimmed = line.trim()
                if (trimmed.isEmpty()) line else publish(resolve(origin, trimmed) ?: trimmed)
            }
        }
    }

    private fun rewriteAttributes(line: String, base: URI): String {
        val regex = Regex("""URI="([^"]*)"""")
        return regex.replace(line) { match ->
            val uri = match.groupValues[1]
            val target = resolve(base, uri) ?: return@replace match.value
            """URI="${publish(target)}""""
        }
    }

    private fun resolve(base: URI, reference: String): String? {
        return try {
            base.resolve(reference).toString()
        } catch (_: Exception) {
            null
        }
    }
}

File: apps/ios/TVM/TVMLiveReflector.swift
import Foundation

/**
 Live TV on the phone goes through the same kind of proxy as desktop Core:
 the player is handed `/api/live/proxy/<token>`, never the provider address.
 */
final class TVMLiveReflector: @unchecked Sendable {
    private var targets: [String: URL] = [:]
    private let lock = NSLock()
    private let session: URLSession

    init(session: URLSession) {
        self.session = session
    }

    func publish(_ url: URL) -> String {
        let token = UUID().uuidString.replacingOccurrences(of: "-", with: "").lowercased()
        lock.lock()
        targets[token] = url
        lock.unlock()
        return "/api/live/proxy/\(token)"
    }

    func serve(_ token: String, method: String) async -> HTTPReply {
        lock.lock()
        let upstream = targets[token]
        lock.unlock()
        guard let upstream else { return .json(404, ["error": "not_found"]) }
        var request = URLRequest(url: upstream)
        request.httpMethod = method == "HEAD" ? "HEAD" : "GET"
        request.timeoutInterval = 20
        request.setValue("VLC/3.0.21 LibVLC/3.0.21", forHTTPHeaderField: "User-Agent")
        do {
            let (data, response) = try await session.data(for: request)
            guard let http = response as? HTTPURLResponse else {
                return .json(502, ["error": "upstream-empty"])
            }
            if !(200..<300).contains(http.statusCode) && http.statusCode != 206 {
                return .json(502, ["error": "upstream-\(http.statusCode)"])
            }
            let type = http.value(forHTTPHeaderField: "Content-Type") ?? ""
            if looksLikePlaylist(upstream, type: type, body: data) {
                let text = String(data: data, encoding: .utf8) ?? ""
                let rewritten = rewrite(text, base: upstream)
                return .bytes(200, type: "application/vnd.apple.mpegurl", data: Data(rewritten.utf8))
            }
            let mediaType = mediaType(upstream, declared: type)
            if method == "HEAD" {
                return HTTPReply(status: 200, headers: ["Content-Type": mediaType], body: Data())
            }
            return .bytes(200, type: mediaType, data: data)
        } catch {
            return .json(502, ["error": "unreachable"])
        }
    }

    private func looksLikePlaylist(_ url: URL, type: String, body: Data) -> Bool {
        if type.lowercased().contains("mpegurl") { return true }
        if url.pathExtension.lowercased() == "m3u8" { return true }
        guard let first = body.first else { return false }
        return first == 0x23 // '#'
    }

    private func mediaType(_ url: URL, declared: String) -> String {
        let mime = declared.split(separator: ";").first.map(String.init)?.trimmingCharacters(in: .whitespaces) ?? ""
        if mime.lowercased().contains("mp2t") { return "video/mp2t" }
        if !mime.isEmpty && !mime.lowercased().contains("octet-stream") { return mime }
        let ext = url.pathExtension.lowercased()
        if ext == "ts" || ext == "m2ts" { return "video/mp2t" }
        if ext == "m3u8" { return "application/vnd.apple.mpegurl" }
        if ext == "mp4" || ext == "m4s" { return "video/mp4" }
        return "video/mp4"
    }

    private func rewrite(_ text: String, base: URL) -> String {
        var lines: [String] = []
        for raw in text.split(separator: "\n", omittingEmptySubsequences: false) {
            var line = String(raw)
            if line.hasPrefix("#") {
                lines.append(rewriteAttributes(line, base: base))
                continue
            }
            let trimmed = line.trimmingCharacters(in: .whitespaces)
            if trimmed.isEmpty {
                lines.append(line)
                continue
            }
            if let target = URL(string: trimmed, relativeTo: base)?.absoluteURL {
                line = publish(target)
            }
            lines.append(line)
        }
        return lines.joined(separator: "\n")
    }

    private func rewriteAttributes(_ line: String, base: URL) -> String {
        guard let regex = try? NSRegularExpression(pattern: "URI=\"([^\"]*)\"") else { return line }
        let ns = line as NSString
        var out = line
        let matches = regex.matches(in: line, range: NSRange(location: 0, length: ns.length)).reversed()
        for match in matches {
            guard match.numberOfRanges >= 2 else { continue }
            let uri = ns.substring(with: match.range(at: 1))
            guard let target = URL(string: uri, relativeTo: base)?.absoluteURL else { continue }
            let replacement = "URI=\"\(publish(target))\""
            out = (out as NSString).replacingCharacters(in: match.range, with: replacement)
        }
        return out
    }
}
